std/algorithm/iteration.d

```diff
@@ -3095,6 +3116,41 @@ if (isInputRange!RoR && isInputRange!(ElementType!RoR)
                 return copy;
             }
         }
+
+        static if (isBidirectional)
+        {
+            void popBack()
+            {
+                assert(!_items.empty, "Attempting to popBack an empty joiner.");
+
+                if (!_currentSep.empty)
+                {
+                    _currentSep.popBack();
```

Comment:
This code is not covered and it will probably fail to compile for situations where the separator range is not a random access range as _currentSep will not be a bidirectional range in that case. Take a look at https://github.com/dlang/phobos/pull/8185/files#diff-b13c31c11ad473d83762bcb8686d251f301236020f8751812580edcec49ef4c5R2963 .

Make sure that all the lines are covered by tests.

Reply:
> Just to be sure, this is the desired behaviour for joiner with separator too, right?

What are you asking more specifically? I don't understand the question.

> Please consider this example of the joiner without separator :

Yes, that example is correct.

Reply:
The behaviour for joiner with separator should be like this, right?
```d
import std.algorithm.iteration : joiner;
import std.stdio : printf, writeln;

void main()
{
    auto r = joiner(["abc", "def", "ghi"],"|?");
    r.popBack;
    auto rCopy = r.save();
    while(!r.empty)
    {
        printf("%c",r.back);
        r.popBack;
    } //prints hg?|fed?|cba
    writeln();
    while(!rCopy.empty)
    {
        printf("%c",rCopy.front);
        rCopy.popFront;
    } //prints abc|?def|?ghi
    writeln();
}
```

Reply:
No, I don't think that the separator should be inserted backwards. It doesn't really make sense  to that. If the user wants to also reverse the separator it can use "|?".retro. Anyway, this should be clarified in the docs. Come to think about it, we can leave the separator to be a forward range as it currently is.

Reply:
Oh, good point, thanks!